loom/engine/bindings/loom/lmSQLite.cpp

```diff
@@ -0,0 +1,384 @@
+        }
+        return result;
+    }
+
+    int columnType(int col)
+    {
+        return sqlite3_column_type(statementHandle, col);
+    }
+
+    int columnInt(int col)
+    {
+        return sqlite3_column_int(statementHandle, col);
+    }
+
+    double columnDouble(int col)
+    {
+        return sqlite3_column_double(statementHandle, col);
+    }
+
+    const char* columnString(int col)
+    {
+        const char *text = (const char *)sqlite3_column_text(statementHandle, col);
+        if(text == NULL)
+        {
+            return NULL;
+        }
+
+        //copy the string into our static return array so it survives the trip to Loomscript Land!
+        strcpy(szReturnString, text);
+        return szReturnString;
+    }
+
+    int columnBytes(lua_State *L)
+    {
+        //get the column
+        int col = (int)lua_tonumber(L, 1);
+
+        //get the blob from the column
+        void *blob = (void *)sqlite3_column_blob(statementHandle, col);
+        if(blob == NULL)
+        {
+            lua_pushnil(L);
+            return 1;
+        }
+        int size = sqlite3_column_bytes(statementHandle, col);
+
+        //valid blob so allocate byte array for it
+        utByteArray *bytes = new utByteArray();
+        bytes->allocateAndCopy(blob, size);
+
+        //go lua go!
+        lualoom_pushnative<utByteArray>(L, bytes);
+        return 1;
+    }
+
+    int reset()
+    {
+        int result = sqlite3_reset(statementHandle); 
+        if(result != SQLITE_OK)
+        {
+            lmLogError(gSQLiteGroup, "Error calling reset for database: %s with Result Code: %i", parentDB->getDBName(), result);
+        }
+        return result;        
+    }
+
+    int finalize()
+    {
+        int result = sqlite3_finalize(statementHandle); 
+        if(result != SQLITE_OK)
+        {
+            lmLogError(gSQLiteGroup, "Error calling finalize for database: %s with Result Code: %i", parentDB->getDBName(), result);
+        }
+        return result;        
+    }
+};
+
+char Statement::szReturnString[4096] = "";
+
+
+
+
+//**Connection** external function initialisation
+Statement *Connection::prepare(const char *query)
+{
+    Statement *s = new Statement(this);
+
+    //prepare the database with the query provided
+    int res = sqlite3_prepare_v2(dbHandle, query, -1, &s->statementHandle, NULL);
+    if(res != SQLITE_OK)
+    {
+        lmLogError(gSQLiteGroup, "Error preparing the SQLite database: %s with message: %s", szDBName, getErrorMessage());
+    }
+    return s;
+}
+Connection *Connection::open(const char *database, int flags)
+{
+    Connection *c = new Connection();
+
+    //prefix the system writable path in front of the database name
+    strcpy(c->szDBName, database);
+    sprintf(c->szDBFullPath, "%s%s", cocos2d::CCFileUtils::sharedFileUtils()->getWriteablePath().c_str(), c->szDBName);
```

Comment:
Cool, utString sounds like a better option for sure.

Not sure about how do deal with paths. Thoughts?  Add another param to
::open that allows abs path to db, if null, use code already in cpp to
prepend system writable?

Have to somehow guarantee that db ends up in system writable location.
On 12 Feb 2015 5:35 AM, "Ben Garney" <notifications@github.com> wrote:

> In loom/engine/bindings/loom/lmSQLite.cpp
> <https://github.com/LoomSDK/LoomSDK/pull/261#discussion_r24556292>:
>
> > +
> > +    //prepare the database with the query provided
> > +    int res = sqlite3_prepare_v2(dbHandle, query, -1, &s->statementHandle, NULL);
> > +    if(res != SQLITE_OK)
> > +    {
> > +        lmLogError(gSQLiteGroup, "Error preparing the SQLite database: %s with message: %s", szDBName, getErrorMessage());
> > +    }
> > +    return s;
> > +}
> > +Connection *Connection::open(const char *database, int flags)
> > +{
> > +    Connection *c = new Connection();
> > +
> > +    //prefix the system writable path in front of the database name
> > +    strcpy(c->szDBName, database);
> > +    sprintf(c->szDBFullPath, "%s%s", cocos2d::CCFileUtils::sharedFileUtils()->getWriteablePath().c_str(), c->szDBName);
>
> What happens if it's an absolute path? Also you have fixed buffers but no
> bounds checking. Is utString a better choice here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LoomSDK/LoomSDK/pull/261/files#r24556292>.
>